lib/experimental/class-wp-html-processor.php

```diff
@@ -0,0 +1,720 @@
+		$existing = $this->scanner->attributes[ $name ];
+
+		$this->modifications->replacements[] = array(
+			$existing->attribute_start,
+			$existing->attribute_end,
+			self::serialize_attribute( $name, $value )
+		);
+	}
+
+
+	public function remove_class( $class_name ) {
+		$this->modifications->class_changes[ $class_name ] = false;
+	}
+
+
+	public function remove_attribute( $name ) {
+		if ( ! array_key_exists( $name, $this->scanner->attributes ) ) {
+			return $this;
+		}
+
+		/** @var WP_HTML_Attribute_Token $existing */
+		$existing = $this->scanner->attributes[ $name ];
+
+		$this->modifications->replacements[] = array(
+			$existing->attribute_start,
+			$existing->attribute_end,
+			''
+		);
+	}
+
+
+	public static function serialize_attribute( $name, $value ) {
+		$value = str_replace( '"', "&quot;", $value );
+
+		return "{$name}=\"{$value}\"";
+	}
+
+
+	public function commit_class_changes() {
+		if ( empty( $this->modifications->class_changes ) ) {
+			return;
+		}
+
+		$existing_class = array_key_exists( 'class', $this->scanner->attributes )
+			? substr(
+				$this->scanner->document,
+				$this->scanner->attributes['class']->value->start,
+				$this->scanner->attributes['class']->value->length
+			)
+			: '';
+
+		// Remove unwanted classes.
+		$new_class = preg_replace_callback(
+			'~(?:^|[ \t])([^ \t]+)~miu',
+			function ( $matches ) {
+				list( $full_match, $class_name ) = $matches;
+
+				$comparable_name = WP_Tag_Find_Descriptor::comparable( $class_name );
+				if (
+					 array_key_exists( $comparable_name, $this->modifications->class_changes ) &&
+					 false === $this->modifications->class_changes[ $comparable_name ]
+				) {
+					 return '';
+				}
+
+				return $full_match;
+			},
+			$existing_class
+		);
+
+		// Add new classes.
+		foreach ( $this->modifications->class_changes as $class => $should_include ) {
+			if ( $should_include ) {
+				$new_class .= " {$class}";
+			}
+		}
+
+		$this->set_attribute( 'class', trim( $new_class ) );
+		$this->modifications->class_changes = array();
+	}
+
+
+	public function apply() {
+		$document = $this->scanner->document;
+
+		$this->commit_class_changes();
+
+		/*
+		 * Since we could be asking to modify attributes in a different
+		 * order than where they appear in the HTMl stream we need to
+		 * sort the modification replacements so we can progress
+		 * linearly and in a single pass.
+		 */
+		usort( $this->modifications->replacements, static function ( $a, $b ) {
+			if ( $a[0] === $b[0] ) {
+				return 0;
+			}
+
+			return $a[0] < $b[0] ? -1 : 1;
+		} );
+
+		$output = '';
+		$at = 0;
+		foreach ( $this->modifications->replacements as $replacement ) {
+			list( $start, $end, $text ) = $replacement;
+
+			if ( $start > $at ) {
+				$output .= substr( $document, $at, $start - $at );
+			}
+
+			$output .= $text;
+			$at = $end;
+		}
+
+		$output .= substr( $document, $at );
+		return $output;
+	}
+
+
+	public function __toString() {
+		return $this->apply();
```

Comment:
I'd cache this result. From there, we could do two things:

* Check if there were any modifications requested since the last caching
* Introduce a "locked"/"finished" state in which you cannot apply any further updates anymore

Reply:
not something I want to do yet for the sake of simplicity of the code, but definitely something I thought about and planned on introducing at some point (if it still makes sense then)